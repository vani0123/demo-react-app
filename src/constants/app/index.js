/**
 * The file contents for the application constants.
 * All the global or application level constants must be place here.
 */

import HTTP_CONSTANTS from '../http';

const APP_CONSTANTS = {
  /**
    Defines constant value for different environment type that will be used
    by config/environment/index.js
  */
  ENVIRONMENT: {
    DEV: 'dev',
    PROD: 'prod',
    QA: 'qa',
    STAGE: 'stage'
  },
  /**
    Defines constant value for route type that will be used for types in
    route configuration routes/app.js
  */
  ROUTE_TYPES: {
    PUBLIC: 'publicRoute'
  },
  /**
    Defines constant value for theme type that will be used by currentTheme redux state.
  */
  THEME: {
    DARK: 'DARK',
    DEFAULT: 'DEFAULT'
  },
  /**
   * Defines constant value for logger that will be used by logging service and utils.
   */
  LOGGER: {
    /**
     * Defines constant value for logger event that will used by request and
     * response interception to log http event info.
     */
    EVENT: {
      HTTP_REQUEST: 'onHttpRequest',
      HTTP_RESPONSE: 'onHttpResponse'
    },
    /**
     * Defines constant value for log level for logger.
     */
    LEVEL: {
      INFO: 'info',
      WARNING: 'warning',
      ERROR: 'error'
    },
    /**
     * Defines constant value for whitelisted http statuses as INFO for logger.
     * Every other error statuses will be considered as errors by logger.
     */
    INFO_STATUSES: [HTTP_CONSTANTS.STATUS_CODES.UNAUTHORIZED]
  },
  /**
    Defines constant value for http method that will be used by http service.
  */
  HTTP: {
    METHOD: {
      GET: 'GET',
      POST: 'POST',
      PUT: 'PUT',
      DELETE: 'DELETE',
      PATCH: 'PATCH'
    }
  },
  /**
    Defines constant value for error type that will be used by error boundary
    for error classification.
  */
  ERROR_TYPE: {
    BackendError: 'BackendError',
    FrontendError: 'FrontendError',
    UnknownError: 'UnknownError'
  }
};

export default APP_CONSTANTS;
