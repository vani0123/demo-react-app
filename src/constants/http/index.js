/**
 * The file contents for the http constants.
 * All the HTTP, NETWORK related constants must be place here.
 */

const HTTP_CONSTANTS = {
  /**
    Defines constant value for http status code.
  */
  STATUS_CODES: {
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    METHOD_NOT_ALLOWED: 405,
    REQUEST_TIMEOUT: 408,
    CONFLICT: 409,
    UNPROCESSABLE_ENTITY: 422,
    INTERNAL_SERVER_ERROR: 500,
    BAD_GATEWAY: 502,
    SERVICE_UNAVAILABLE: 503,
    GATEWAY_TIMEOUT: 504,
    OK: 200
  },
  /**
    Defines constant value for specific error code that will be used by rest apis
  */
  ERROR_CODES: {
    NOT_AUTHORIZED: 'NotAuthorizedException'
  },
  /**
    Defines constant value for http method that will be used by http service.
  */
  REQUEST_TYPE: {
    GET: 'get',
    POST: 'post',
    PUT: 'put',
    DELETE: 'delete',
    PATCH: 'patch'
  }
};

export default HTTP_CONSTANTS;
