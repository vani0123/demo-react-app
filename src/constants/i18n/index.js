/**
 * The file contents for the i18n constants.
 * All the I18N related constants must be place here.
 */

const I18N_CONSTANTS = {
  /**
    Defines constant value for i18n namespaces.
    These namespace will categories translations for different application features.
  */
  NAMESPACE: {
    APP: 'app',
    HOME: 'home'
  },
  /**
    Defines constant value for whitelisted namespace for i18n.
  */
  WHITE_LIST_NAMESPACE: ['app', 'home'],
  /**
    Defines constant value for whitelisted language for i18n.
  */
  WHITE_LIST: ['en-US'],
  /**
    Defines constant value for default namespace for i18n.
  */
  DEFAULT_NS: 'app',
  /**
    Defines constant value for default language for i18n.
  */
  DEFAULT_LANGUAGE: 'en-US',
  /**
    Defines constant value for fallback language for i18n.
  */
  FALLBACK_LANGUAGE: 'en-US',
  /**
    Defines constant value for translation request for i18n.
  */
  LOOKUP: 'translations/{{ns}}/{{lng}}.json'
};

export default I18N_CONSTANTS;
