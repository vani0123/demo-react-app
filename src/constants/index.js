/**
 * Main export for constants file that will be used across application.
 *
 * eg -
 * import { APP_CONSTANTS } from './constants';
 * import * as CONSTANTS from './constants';
 */

import APP_CONSTANTS from './app';
import ROUTE_CONSTANTS from './route';
import I18N_CONSTANTS from './i18n';
import THEME_CONSTANTS from './theme';
import HTTP_CONSTANTS from './http';

export {
  APP_CONSTANTS,
  ROUTE_CONSTANTS,
  I18N_CONSTANTS,
  THEME_CONSTANTS,
  HTTP_CONSTANTS
};
