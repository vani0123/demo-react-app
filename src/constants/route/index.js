/**
 * The file contents for the route constants.
 * All the route related constants must be place here.
 */
const ROUTE_CONSTANTS = {
  /**
    Defines constant value application root url.
  */
  ROOT: '/',
  WILDCARDS: {
    INTERNALSERVERERROR: '/internal-server-error',
    NOTFOUND: '/not-found',
    UNAUTHORIZED: '/unauthorized'
  }
};
export default ROUTE_CONSTANTS;
