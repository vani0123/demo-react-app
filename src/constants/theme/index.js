/**
 * The file contents for the theme constants.
 * All the theme style guide related constants must be place here.
 */

import THEME_DEFAULT from './default/default';
import THEME_DARK from './dark/dark';

const THEME_CONSTANTS = {
  DEFAULT: THEME_DEFAULT,
  DARK: THEME_DARK
};

export default THEME_CONSTANTS;
