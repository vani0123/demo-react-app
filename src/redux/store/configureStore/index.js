import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import createSagaMiddleware from 'redux-saga';
import rootReducer from '../../reducers';
import { initSagas } from '../../sagas/initSagas';

export default function configureStore(initialState) {
  const shouldHideDevTool = process.env.NODE_ENV === 'production'
    || typeof document === 'undefined'
    || !window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

  const composeEnhancers = shouldHideDevTool
    ? compose
    : window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__; // add support for Redux dev tools

  const sagaMiddleware = createSagaMiddleware({
    onError(error) {
      setImmediate(() => {
        throw error;
      });
    }
  });
  const store = createStore(
    rootReducer,
    initialState,
    composeEnhancers(
      applyMiddleware(sagaMiddleware, reduxImmutableStateInvariant())
    )
  );

  initSagas(sagaMiddleware);

  return store;
}
