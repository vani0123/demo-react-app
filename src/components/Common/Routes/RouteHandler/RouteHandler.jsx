import React from 'react';
import { PropTypes } from 'prop-types';
import { PublicRoute } from '..';
import { APP_CONSTANTS } from '../../../../constants';

/**
 * RouteHandler component, responsible for rendering route wrapped with respective guards as defined in route config.
 * This handler pass the target component to the route guard based on the type property defined in
 * route configuration 'src/routes/app.js'.
 *
 * This behavior can be overridden by adding a specific switch case in the below switch statement
 * to handle specific/new route guard.
 *
 * eg -
 * For authenticated routes, add another case statement in below switch statement that will allow route config object
 * with type property sets to Authenticated and pass target component to Authenticated Guard.
 *
 * These guard are high ordered component that are wrapper over the route that will be rendered by route handler.
 *
 * @example
 * const route = {
 *   path: '/example',
 *   component: Home,
 *   exact: true,
 *   type: 'Public',
 * };
 *
 * <RouteHandler { ...route } />
 */
const RouteHandler = (props) => {
  const { type } = props;

  let Handler;

  switch (type) {
    case APP_CONSTANTS.ROUTE_TYPES.PUBLIC:
      Handler = PublicRoute;
      break;
    default:
      Handler = PublicRoute;
  }

  return <Handler {...props} />;
};

/**
 * Default PropTypes definition for PublicRoute component.
 *
 * routeAccess - Defines the access control property in route config.
 */
RouteHandler.defaultProps = {
  routeAccess: null
};

/**
 * Default PropTypes definition for PublicRoute component.
 *
 * type - Defines the type of route defined in route configuration that will be rendered by RouteHandler.
 * routeAccess - Defines the access control property in route config.
 * path: Defines the url to which target component will be mapped.
 */
RouteHandler.propTypes = {
  type: PropTypes.string.isRequired,
  routeAccess: PropTypes.shape({}),
  path: PropTypes.string.isRequired
};

export default RouteHandler;
