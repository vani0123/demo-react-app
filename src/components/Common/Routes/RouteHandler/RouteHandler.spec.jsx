import React from 'react';
import { mount } from 'enzyme';
import { HashRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import configureStore from '../../../../redux/store/configureStore';
import { APP_CONSTANTS } from '../../../../constants';

import RouteHandler from './RouteHandler';
import { Home } from '../../..';

const setup = (props) => {
  const store = configureStore();

  const defaultProps = {
    component: Home,
    type: APP_CONSTANTS.ROUTE_TYPES.PUBLIC,
    location: { pathname: '/home' },
    path: '/home',
    ...props
  };

  return mount(
    <Router>
      <Provider store={store}>
        <RouteHandler {...defaultProps} />
      </Provider>
    </Router>
  );
};

describe('The route handler', () => {
  it('should renders without crashing', () => {
    const wrapper = setup(true);

    expect(wrapper.find('PublicRoute').exists());
    expect(wrapper.find('Route').exists()).toBe(true);
  });

  it('should have valid props for public types', () => {
    const wrapper = setup(false, {
      type: APP_CONSTANTS.ROUTE_TYPES.PUBLIC,
      routes: {}
    });

    const { type } = wrapper.find('RouteHandler').props();

    expect(wrapper.find('PublicRoute').exists()).toBe(true);
    expect(type).toBe('publicRoute');
  });

  it('should have valid snapshot', () => {
    const wrapper = setup(true);

    expect(wrapper).toMatchSnapshot();
  });
});
