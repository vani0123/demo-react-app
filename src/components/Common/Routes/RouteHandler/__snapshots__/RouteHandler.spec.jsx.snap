// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The route handler should have valid snapshot 1`] = `ReactWrapper {}`;
